File: NomadTracker/Services/RulesEngine.swift
import Foundation

/// Calculates visa day counts based on jurisdiction rules
struct RulesEngine {

    private let calendar = Calendar.current

    // MARK: - Core Calculations

    /// Calculate how many days have been used in a jurisdiction
    func daysUsed(for jurisdiction: Jurisdiction, records: [StayRecord], asOf date: Date = .now) -> Int {
        let today = calendar.startOfDay(for: date)
        let relevantRecords = records.filter { $0.jurisdictionId == jurisdiction.id }

        switch jurisdiction.ruleType {
        case .rolling(_, let windowDays):
            return daysInRollingWindow(records: relevantRecords, windowDays: windowDays, asOf: today)

        case .perVisit:
            return daysInCurrentVisit(records: relevantRecords, asOf: today)

        case .calendarYear:
            return daysInCalendarYear(records: relevantRecords, asOf: today)
        }
    }

    /// Calculate how many days remain in a jurisdiction
    func daysRemaining(for jurisdiction: Jurisdiction, records: [StayRecord], asOf date: Date = .now) -> Int {
        let used = daysUsed(for: jurisdiction, records: records, asOf: date)
        return max(0, jurisdiction.ruleType.maxDays - used)
    }

    /// Calculate the percentage of days used (0.0 to 1.0)
    func usagePercentage(for jurisdiction: Jurisdiction, records: [StayRecord], asOf date: Date = .now) -> Double {
        let used = daysUsed(for: jurisdiction, records: records, asOf: date)
        let max = jurisdiction.ruleType.maxDays
        guard max > 0 else { return 0 }
        return min(1.0, Double(used) / Double(max))
    }

    /// Calculate the date when you must leave (if staying continuously from today)
    func mustLeaveBy(for jurisdiction: Jurisdiction, records: [StayRecord], asOf date: Date = .now) -> Date? {
        let stayableDays = continuousStayDaysAvailable(for: jurisdiction, records: records, asOf: date)
        let today = calendar.startOfDay(for: date)

        guard stayableDays > 0 else { return today }
        return calendar.date(byAdding: .day, value: stayableDays - 1, to: today)
    }

    /// Calculate how many continuous days you can legally stay starting today.
    func continuousStayDaysAvailable(
        for jurisdiction: Jurisdiction,
        records: [StayRecord],
        asOf date: Date = .now
    ) -> Int {
        let today = calendar.startOfDay(for: date)
        let relevantRecords = records.filter { $0.jurisdictionId == jurisdiction.id }
        let isPresentToday = relevantRecords.contains {
            calendar.isDate($0.date, inSameDayAs: today)
        }

        switch jurisdiction.ruleType {
        case .rolling(let maxDays, let windowDays):
            return projectedRollingStayDays(
                records: relevantRecords,
                maxDays: maxDays,
                windowDays: windowDays,
                asOf: today
            )

        case .perVisit(let maxDays):
            let used = daysInCurrentVisit(records: relevantRecords, asOf: today)
            let inclusiveDays = isPresentToday ? (maxDays - used + 1) : (maxDays - used)
            return max(0, inclusiveDays)

        case .calendarYear(let maxDays):
            let used = daysInCalendarYear(records: relevantRecords, asOf: today)
            let inclusiveDays = isPresentToday ? (maxDays - used + 1) : (maxDays - used)
            return max(0, inclusiveDays)
        }
    }

    /// For rolling windows, quantify how much older days extend a continuous stay.
    func projectedExtraDaysFromWindowExpiry(
        for jurisdiction: Jurisdiction,
        records: [StayRecord],
        asOf date: Date = .now
    ) -> Int {
        guard case .rolling = jurisdiction.ruleType else { return 0 }

        let today = calendar.startOfDay(for: date)
        let relevantRecords = records.filter { $0.jurisdictionId == jurisdiction.id }
        let isPresentToday = relevantRecords.contains {
            calendar.isDate($0.date, inSameDayAs: today)
        }
        let naiveDays = max(0, daysRemaining(for: jurisdiction, records: records, asOf: today) + (isPresentToday ? 1 : 0))
        let projectedDays = continuousStayDaysAvailable(for: jurisdiction, records: records, asOf: today)

        return max(0, projectedDays - naiveDays)
    }

    /// If you leave today, when would this jurisdiction return to a full allowance?
    func fullAllowanceResetDate(
        for jurisdiction: Jurisdiction,
        records: [StayRecord],
        asOf date: Date = .now
    ) -> Date? {
        let today = calendar.startOfDay(for: date)
        let relevantRecords = records.filter { $0.jurisdictionId == jurisdiction.id }

        switch jurisdiction.ruleType {
        case .rolling(_, let windowDays):
            let inWindowDates = relevantRecords
                .map { calendar.startOfDay(for: $0.date) }
                .filter { recordedDay in
                    guard let windowStart = calendar.date(byAdding: .day, value: -(windowDays - 1), to: today) else {
                        return false
                    }
                    return recordedDay >= windowStart && recordedDay <= today
                }

            guard let latestRelevantDate = inWindowDates.max() else { return nil }
            return calendar.date(byAdding: .day, value: windowDays, to: latestRelevantDate)

        case .calendarYear:
            guard daysInCalendarYear(records: relevantRecords, asOf: today) > 0 else { return nil }
            return calendarYearReset(asOf: today)

        case .perVisit:
            return nil
        }
    }

    /// For rolling windows: find the date when days start "falling off"
    func nextDayFallsOff(for jurisdiction: Jurisdiction, records: [StayRecord], asOf date: Date = .now) -> (date: Date, count: Int)? {
        guard case .rolling(_, let windowDays) = jurisdiction.ruleType else { return nil }

        let today = calendar.startOfDay(for: date)
        let relevantRecords = records
            .filter { $0.jurisdictionId == jurisdiction.id }
            .sorted { $0.date < $1.date }

        // Find the earliest record that's still within the window
        guard let windowStart = calendar.date(byAdding: .day, value: -(windowDays - 1), to: today) else { return nil }
        let inWindowRecords = relevantRecords.filter { $0.date >= windowStart && $0.date <= today }

        guard let earliestInWindow = inWindowRecords.first else { return nil }

        // The earliest day falls off when it's windowDays after its date
        if let fallOffDate = calendar.date(byAdding: .day, value: windowDays, to: earliestInWindow.date) {
            // Count how many days fall off on that same date
            let count = inWindowRecords.filter {
                calendar.isDate($0.date, inSameDayAs: earliestInWindow.date)
            }.count
            return (fallOffDate, count)
        }
        return nil
    }

    /// For calendar year: when does the counter reset?
    func calendarYearReset(asOf date: Date = .now) -> Date? {
        let year = calendar.component(.year, from: date)
        return calendar.date(from: DateComponents(year: year + 1, month: 1, day: 1))
    }

    // MARK: - Status

    /// Get the urgency level for display purposes
    func urgencyLevel(for jurisdiction: Jurisdiction, records: [StayRecord], asOf date: Date = .now) -> UrgencyLevel {
        let remaining = daysRemaining(for: jurisdiction, records: records, asOf: date)
        let max = jurisdiction.ruleType.maxDays
        let used = daysUsed(for: jurisdiction, records: records, asOf: date)
        let today = calendar.startOfDay(for: date)
        let isPresentToday = records.contains {
            $0.jurisdictionId == jurisdiction.id && calendar.isDate($0.date, inSameDayAs: today)
        }

        // If they haven't used any days, it's safe
        if used == 0 { return .safe }

        let percentRemaining = Double(remaining) / Double(max)

        if used > max { return .expired }
        if remaining == 0 { return isPresentToday ? .critical : .expired }
        if remaining <= 7 { return .critical }
        if percentRemaining < 0.15 { return .critical }
        if percentRemaining < 0.33 { return .warning }
        if percentRemaining < 0.66 { return .caution }
        return .safe
    }

    // MARK: - Private Helpers

    /// Count days within a rolling window
    private func daysInRollingWindow(records: [StayRecord], windowDays: Int, asOf date: Date) -> Int {
        guard let windowStart = calendar.date(byAdding: .day, value: -(windowDays - 1), to: date) else { return 0 }

        // Count unique days in the window
        let uniqueDays = Set(records
            .filter { $0.date >= windowStart && $0.date <= date }
            .map { calendar.startOfDay(for: $0.date) }
        )
        return uniqueDays.count
    }

    /// Count days in the current continuous visit
    private func daysInCurrentVisit(records: [StayRecord], asOf date: Date) -> Int {
        let sortedDates = Set(records.map { calendar.startOfDay(for: $0.date) }).sorted(by: >)

        var count = 0
        var expectedDate = date

        for recordDate in sortedDates {
            let dayStart = calendar.startOfDay(for: expectedDate)
            if calendar.isDate(recordDate, inSameDayAs: dayStart) {
                count += 1
                guard let prevDay = calendar.date(byAdding: .day, value: -1, to: expectedDate) else { break }
                expectedDate = prevDay
            } else {
                break
            }
        }
        return count
    }

    /// Count days in the current calendar year
    private func daysInCalendarYear(records: [StayRecord], asOf date: Date) -> Int {
        let year = calendar.component(.year, from: date)
        guard let yearStart = calendar.date(from: DateComponents(year: year, month: 1, day: 1)) else { return 0 }

        let uniqueDays = Set(records
            .filter { $0.date >= yearStart && $0.date <= date }
            .map { calendar.startOfDay(for: $0.date) }
        )
        return uniqueDays.count
    }

    /// Simulate a continuous stay to find the last legal day in a rolling window.
    private func projectedRollingStayDays(
        records: [StayRecord],
        maxDays: Int,
        windowDays: Int,
        asOf date: Date
    ) -> Int {
        let existingDays = Set(records.map { calendar.startOfDay(for: $0.date) })
        var occupiedDays = existingDays
        var projectedDate = date
        var legalStayDays = 0

        while legalStayDays < maxDays {
            occupiedDays.insert(projectedDate)

            guard let windowStart = calendar.date(byAdding: .day, value: -(windowDays - 1), to: projectedDate) else {
                break
            }

            let usedDays = occupiedDays.filter { $0 >= windowStart && $0 <= projectedDate }.count
            if usedDays > maxDays {
                break
            }

            legalStayDays += 1

            guard let nextDate = calendar.date(byAdding: .day, value: 1, to: projectedDate) else {
                break
            }
            projectedDate = nextDate
        }

        return legalStayDays
    }
}

/// How urgent the visa situation is
enum UrgencyLevel {
    case safe       // > 66% remaining
    case caution    // 33-66% remaining
    case warning    // 15-33% remaining
    case critical   // < 15% or < 7 days remaining
    case expired    // 0 days remaining
}

File: docs/js/services/rules-engine.js
// Visa day counting algorithms — ported from RulesEngine.swift
import { RULE_TYPES } from '../data/jurisdictions.js';
import { toDateStr, todayStr, parseDate, addDays, diffDays, getRecords } from './storage.js';

// Core: how many days used in a jurisdiction
export function daysUsed(jurisdiction, records, asOf = todayStr()) {
  const relevant = records.filter(r => r.jurisdictionId === jurisdiction.id);

  switch (jurisdiction.ruleType) {
    case RULE_TYPES.ROLLING:
      return daysInRollingWindow(relevant, jurisdiction.windowDays, asOf);
    case RULE_TYPES.PER_VISIT:
      return daysInCurrentVisit(relevant, asOf);
    case RULE_TYPES.CALENDAR_YEAR:
      return daysInCalendarYear(relevant, asOf);
    default:
      return 0;
  }
}

// Core: how many days remain
export function daysRemaining(jurisdiction, records, asOf = todayStr()) {
  const used = daysUsed(jurisdiction, records, asOf);
  return Math.max(0, jurisdiction.maxDays - used);
}

// Usage percentage (0-1)
export function usagePercentage(jurisdiction, records, asOf = todayStr()) {
  const used = daysUsed(jurisdiction, records, asOf);
  if (jurisdiction.maxDays <= 0) return 0;
  return Math.min(1, used / jurisdiction.maxDays);
}

// How many continuous days can you legally stay starting from asOf
export function continuousStayDaysAvailable(jurisdiction, records, asOf = todayStr()) {
  const relevant = records.filter(r => r.jurisdictionId === jurisdiction.id);
  const isPresentToday = relevant.some(r => r.date === asOf);

  switch (jurisdiction.ruleType) {
    case RULE_TYPES.ROLLING:
      return projectedRollingStayDays(relevant, jurisdiction.maxDays, jurisdiction.windowDays, asOf);

    case RULE_TYPES.PER_VISIT: {
      const used = daysInCurrentVisit(relevant, asOf);
      const inclusive = isPresentToday ? (jurisdiction.maxDays - used + 1) : (jurisdiction.maxDays - used);
      return Math.max(0, inclusive);
    }

    case RULE_TYPES.CALENDAR_YEAR: {
      const used = daysInCalendarYear(relevant, asOf);
      const inclusive = isPresentToday ? (jurisdiction.maxDays - used + 1) : (jurisdiction.maxDays - used);
      return Math.max(0, inclusive);
    }

    default:
      return 0;
  }
}

// Date when you must leave (if staying continuously)
export function mustLeaveBy(jurisdiction, records, asOf = todayStr()) {
  const stayable = continuousStayDaysAvailable(jurisdiction, records, asOf);
  if (stayable <= 0) return asOf;
  return addDays(asOf, stayable - 1);
}

// For rolling windows: how many extra days you gain from old days falling off
export function projectedExtraDays(jurisdiction, records, asOf = todayStr()) {
  if (jurisdiction.ruleType !== RULE_TYPES.ROLLING) return 0;

  const relevant = records.filter(r => r.jurisdictionId === jurisdiction.id);
  const isPresentToday = relevant.some(r => r.date === asOf);
  const naiveDays = Math.max(0, daysRemaining(jurisdiction, records, asOf) + (isPresentToday ? 1 : 0));
  const projected = continuousStayDaysAvailable(jurisdiction, records, asOf);

  return Math.max(0, projected - naiveDays);
}

// For rolling windows: when does the earliest day fall off?
export function nextDayFallsOff(jurisdiction, records, asOf = todayStr()) {
  if (jurisdiction.ruleType !== RULE_TYPES.ROLLING) return null;

  const relevant = records
    .filter(r => r.jurisdictionId === jurisdiction.id)
    .sort((a, b) => a.date.localeCompare(b.date));

  const windowStart = addDays(asOf, -jurisdiction.windowDays);
  const inWindow = relevant.filter(r => r.date >= windowStart && r.date <= asOf);
  if (inWindow.length === 0) return null;

  const earliest = inWindow[0];
  const fallOffDate = addDays(earliest.date, jurisdiction.windowDays + 1);
  const count = inWindow.filter(r => r.date === earliest.date).length;

  return { date: fallOffDate, count };
}

// For calendar year: when does the counter reset?
export function calendarYearResetDate(asOf = todayStr()) {
  const year = parseDate(asOf).getFullYear();
  return `${year + 1}-01-01`;
}

// Urgency level for display
export function urgencyLevel(jurisdiction, records, asOf = todayStr()) {
  const remaining = daysRemaining(jurisdiction, records, asOf);
  const used = daysUsed(jurisdiction, records, asOf);
  const max = jurisdiction.maxDays;
  const isPresentToday = records.some(
    r => r.jurisdictionId === jurisdiction.id && r.date === asOf
  );

  if (used === 0) return 'safe';
  if (used > max) return 'expired';
  if (remaining === 0) return isPresentToday ? 'critical' : 'expired';
  if (remaining <= 7) return 'critical';

  const pct = remaining / max;
  if (pct < 0.15) return 'critical';
  if (pct < 0.33) return 'warning';
  if (pct < 0.66) return 'caution';
  return 'safe';
}

// --- Private helpers ---

function daysInRollingWindow(records, windowDays, asOf) {
  const windowStart = addDays(asOf, -(windowDays - 1));
  const uniqueDays = new Set(
    records
      .filter(r => r.date >= windowStart && r.date <= asOf)
      .map(r => r.date)
  );
  return uniqueDays.size;
}

function daysInCurrentVisit(records, asOf) {
  const uniqueDates = [...new Set(records.map(r => r.date))].sort().reverse();
  let count = 0;
  let expected = asOf;

  for (const date of uniqueDates) {
    if (date === expected) {
      count++;
      expected = addDays(expected, -1);
    } else {
      break;
    }
  }
  return count;
}

function daysInCalendarYear(records, asOf) {
  const year = parseDate(asOf).getFullYear();
  const yearStart = `${year}-01-01`;
  const uniqueDays = new Set(
    records
      .filter(r => r.date >= yearStart && r.date <= asOf)
      .map(r => r.date)
  );
  return uniqueDays.size;
}

// Simulate continuous stay forward for rolling windows
function projectedRollingStayDays(records, maxDays, windowDays, asOf) {
  const existingDays = new Set(records.map(r => r.date));
  const occupiedDays = new Set(existingDays);
  let projectedDate = asOf;
  let legalDays = 0;

  while (legalDays < maxDays) {
    occupiedDays.add(projectedDate);

    const windowStart = addDays(projectedDate, -(windowDays - 1));
    let usedInWindow = 0;
    for (const d of occupiedDays) {
      if (d >= windowStart && d <= projectedDate) usedInWindow++;
    }

    if (usedInWindow > maxDays) break;
    legalDays++;
    projectedDate = addDays(projectedDate, 1);
  }

  return legalDays;
}
